lib/OpenQA/WebAPI/Controller/API/V1/Search.pm

```diff
@@ -123,27 +110,55 @@ sub query {
             }
             $last_filename = $filename;
             push(@results, {occurrence => "$basename/$filename", contents => $contents});
+            # For the purposes of the limit, all lines in the same file count as one
+            $cap--;
+            last if $cap < 1;
         }
-        $cap -= scalar @results;
-        last if $cap < 1;
+    }
+    return \@results;
+}
 
-        # Job templates
-        my $last_group = undef;
-        my $like       = {like => "%${keywords}%"};
-        my $templates  = $self->schema->resultset('JobTemplates')
-          ->search({-or => {name => $like, description => $like}}, {limit => $cap});
-        while (my $template = $templates->next) {
-            my $contents = $template->name . "\n" . $template->description;
-            if ($template->group->id == $last_group) {
-                $results[-1]->{contents} .= "\n$contents";
-                next;
-            }
-            $last_group = $template->group->id;
-            push(@results, {occurrence => $template->group->name, contents => $contents});
+sub _search_job_templates {
+    my ($self, $keywords, $limit) = @_;
+
+    my @results;
+    my $last_group = -1;
+    my $like       = {like => "%${keywords}%"};
+    my $templates  = $self->schema->resultset('JobTemplates')
+      ->search({-or => {name => $like, description => $like}}, {limit => $limit});
```

Comment:
I am not familiar with the database operation, but in my test this `{limit => $limit}` in the search does not work well.  @Martchus may know this.

Reply:
What do you mean by "does not work well"? To my understanding the number of rows is limited to the given amount, i.e. `LIMIT <n>` in SQL.

Reply:
I set `search_results_limit = 10`, then there are 11 results.  Ten of them are the test code, and the last one is job template name.  No job template name is shown is expected, is it ?  Maybe my test is wrong ?

Reply:
Nevermind, I double-checked the produced SQL, and the docs... and apparently the `{ limit => }` does nothing, and other existing openQA code where I got inspiration from may well be a no-op, too. Apparently [slice](https://metacpan.org/pod/DBIx::Class::ResultSet#slice) is what we want. I love DBIX.

Reply:
Feel free to give any feedback, I do not mind :)
I checked the other code that use `limit` in query database,  they use `->complex_query`, I am not sure if it is related.

Reply:
there's also `rows` and `offset`

Reply:
The function `complex_query` uses `rows` when `limit` is passed so that's fine (although it would make sense to use `rows` from the beginning). And yes, `rows` in DBIx corresponds to `LIMIT` in SQL. I have no idea why the DBIx author thought it would be a good idea to deviate from SQL here because it only seems to cause confusion.

Reply:
I couldn't find this in the docs, but experimental verification showed that `LIMIT` was not part of the query. I also proposed #3337 to avoid getting rows we don't need.

Reply:
I should clarify, I did see `rows` mentioned as a work-around for obscure databases with no support for limits, but not in reference to `{limit =>}` or the correct way to cap results.

Reply:
Yes, `slice` is the way to go.

Reply:
While `limit` and `offset` might seem like obvious terms to those of us used to open source databases, they are not actually part of standard SQL. In certain commercial databases the concept works completely different. So you can't really fault the `DBIx::Class` authors for choosing more neutral terms, considering the module is database agnostic.

Reply:
@kraih Mh… ok.
@kalikiana The official documentation does not state that `rows` is a workaround: https://metacpan.org/pod/distribution/DBIx-Class/lib/DBIx/Class/ResultSet.pm#rows  
The documentation even recommends using `rows` (and considers using `slice` for some reason ("A less readable way"): https://metacpan.org/pod/distribution/DBIx-Class/lib/DBIx/Class/Manual/FAQ.pod#..-fetch-a-single-(or-topmost)-row?

Reply:
Probably because requesting one row would end up as `->slice(0)` in that specific case. Which is a bit unintuitive.